crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -721,39 +735,50 @@ where
 
         let change_tick = world.increment_change_tick();
 
+        let param_state = self
+            .state
+            .as_mut()
+            .map(|state| &mut state.param)
+            .expect(Self::ERROR_UNINITIALIZED);
         // SAFETY:
         // - The caller has invoked `update_archetype_component_access`, which will panic
         //   if the world does not match.
         // - All world accesses used by `F::Param` have been registered, so the caller
         //   will ensure that there are no data access conflicts.
-        let params = unsafe {
-            F::Param::get_param(
-                self.param_state.as_mut().expect(Self::PARAM_MESSAGE),
-                &self.system_meta,
-                world,
-                change_tick,
-            )
-        };
+        let params =
+            unsafe { F::Param::get_param(param_state, &self.system_meta, world, change_tick) };
         let out = self.func.run(input, params);
         self.system_meta.last_run = change_tick;
         out
     }
 
     #[inline]
     fn apply_deferred(&mut self, world: &mut World) {
-        let param_state = self.param_state.as_mut().expect(Self::PARAM_MESSAGE);
+        let param_state = self
+            .state
+            .as_mut()
+            .map(|state| &mut state.param)
+            .expect(Self::ERROR_UNINITIALIZED);
         F::Param::apply(param_state, &self.system_meta, world);
     }
 
     #[inline]
     fn queue_deferred(&mut self, world: DeferredWorld) {
-        let param_state = self.param_state.as_mut().expect(Self::PARAM_MESSAGE);
+        let param_state = self
+            .state
+            .as_mut()
+            .map(|state| &mut state.param)
+            .expect(Self::ERROR_UNINITIALIZED);
         F::Param::queue(param_state, &self.system_meta, world);
     }
 
     #[inline]
     unsafe fn validate_param_unsafe(&mut self, world: UnsafeWorldCell) -> bool {
-        let param_state = self.param_state.as_ref().expect(Self::PARAM_MESSAGE);
+        let param_state = self
+            .state
+            .as_ref()
+            .map(|state| &state.param)
+            .expect(Self::ERROR_UNINITIALIZED);
```

Comment:
```suggestion
        let param_state = self
            .state
            .as_ref()
            .expect(Self::ERROR_UNINITIALIZED)
            .param;
```
Does this work? If so, it's a bit simpler.
(same comment for the other places this line appears)